### PR TITLE
Add SnowflakeSqlApiOperator operator

### PR DESCRIPTION
**Pull request by utkarsharma2**
Implemented Snowflake SQL API Operator to support multiple SQL statements, the Snowflake SQL API allows submitting multiple SQL statements in a single request, which currently is not possible with `SnowflakeOperator` as it submits multiple statements sequentially.

This Operator currently uses key pair authentication, so you need to provide private key raw content or private key file path in the snowflake connection along with other details. 

Snowflake SQL API key pair  - Authentication https://docs.snowflake.com/en/developer-guide/sql-api/authenticating.html#label-sql-api-authenticating-key-pair

Where can this operator fit in?
         - To execute multiple SQL statements in a single request
         - To execute the SQL statement and to execute standard queries and most DDL and DML statements

### CODE REVIEW

File: airflow/providers/snowflake/hooks/snowflake_sql_api.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+                "query_tag": query_tag,
+            },
+        }
+        response = requests.post(url, json=data, headers=headers, params=params)
+        try:
+            response.raise_for_status()
+        except requests.exceptions.HTTPError as e:  # pragma: no cover
+            raise AirflowException(
+                f"Response: {e.response.content}, " f"Status Code: {e.response.status_code}"
+            )  # pragma: no cover
```
Comment (utkarsharma2) [inline airflow/providers/snowflake/hooks/snowflake_sql_api.py]:
```suggestion
            ) 
```

**Review by phanikumv — COMMENTED** [review]
Please add documentation and example DAG

**Comment by utkarsharma2** [comment]
@phanikumv Added documents and example

**Comment by utkarsharma2** [comment]
@potiuk, please review

**Comment by potiuk** [comment]
> @potiuk please review

I thin it's better to avoid calling specific people. I might be on holidays or focusing on something else. By calling me personally here (when I was not involved before especially)  you are limiting the chances for things to be reviewed (because others might think I will do it). It's better to just ask for review `in general` - or ping people if they have been reviewing things before than asking people directly to do review..

Note to self - we need to upgrade our CODEOWNERS rules actually as they might be misleding and they are wrong in multiple cases.

**Review by potiuk — APPROVED** [review]
LGTM. Maybe @mik-laj -you would like to have a look ?

**Comment by utkarsharma2** [comment]
@potiuk Sure, I'll keep that in mind going forward. Thanks for reviewing :)

**Comment by potiuk** [comment]
If we don't hear from Snowflake team, we are going t merge it soon anyway @eladkal - might be worth merging before the next wave ? But maybe we can drag attention of the Snowflake team to do more review here (there was somoene at Snowflake who looked at the changes before I recall @mik-laj ?

**Review by ephraimbuddy — APPROVED** [review]

File: airflow/providers/snowflake/hooks/sql_api_generate_jwt.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (eladkal) [inline airflow/providers/snowflake/hooks/sql_api_generate_jwt.py:1]: This is not a hook.
it looks more like util and should be in util path

File: airflow/providers/snowflake/provider.yaml [hunk 3]
```diff
@@ -74,6 +74,8 @@ hooks:
   - integration-name: Snowflake
     python-modules:
       - airflow.providers.snowflake.hooks.snowflake
+      - airflow.providers.snowflake.hooks.snowflake_sql_api
+      - airflow.providers.snowflake.hooks.sql_api_generate_jwt
```
Comment (eladkal) [inline airflow/providers/snowflake/provider.yaml:78]: This is not right. it's not a hook.

File: airflow/providers/snowflake/hooks/sql_api_generate_jwt.py (hunk 2 above)
Comment (utkarsharma2) [reply airflow/providers/snowflake/hooks/sql_api_generate_jwt.py:1]: Yes, that makes sense. I'll raise a separate PR to move it. thanks for pointing it out :)

File: airflow/providers/snowflake/operators/snowflake.py [hunk 4]
```diff
@@ -359,3 +368,159 @@ def __init__(
         self.authenticator = authenticator
         self.session_parameters = session_parameters
         self.query_ids: list[str] = []
+
+
+class SnowflakeSqlApiOperator(SnowflakeOperator):
```
Comment (eladkal) [inline airflow/providers/snowflake/operators/snowflake.py:373]: This is not good.
`SnowflakeOperator` is deprecated. We should not extend it's functionality.

This class needs to be refactored around `SQLExecuteQueryOperator`, `BaseOperator` or possibly have a `SnowflakeBaseOperator` class ?

Comment (eladkal) [reply airflow/providers/snowflake/operators/snowflake.py:373]: For reference... followup PR https://github.com/apache/airflow/pull/31751

**Pull request merged**